Updates

hansmbakker opened the pull request:
Hi,

I made some modifications to the Nuimo SDK:
- Switched to new LedMatrix characteristic
- Added builtin LedMatrix support + Busy indicator
- Added new Touch and LongTouch events
- Added methods for reboot to DFU, Fly sensor calibration
- Added heartbeat logic
- Added functions to read hardware version and Nuimo color
- Replaced characteristics and services loading
- Removed deprecated API calls
  - `(await GattDeviceService.FromIdAsync(deviceInformation.Id)).Device`
  - `_bluetoothLeDevice.GetGattService(serviceGuid)`
  - `service.GetCharacteristics(characteristicGuid)[0]`
- Added `DeviceWatcher` functionality to PairedNuimoManager to dynamically get INuimoControllers (see #5 )
- Add `ThrottledGestureEventOccurred` with `ThrottlePeriod` setting: multiple `NuimoGesture.Rotate` events are combined to prevent triggering event handlers too often
- Applied ReSharper code cleanup

hansmbakker committed:
use recent UWP SDK

hansmbakker committed:
- Remove deprecated API calls
  - (await GattDeviceService.FromIdAsync(deviceInformation.Id)).Device
  - _bluetoothLeDevice.GetGattService(serviceGuid)
  - service.GetCharacteristics(characteristicGuid)[0]
- Add DeviceWatcher to dynamically get INuimoControllers
- Add throttled event with ThrottlePeriod setting: multiple NuimoGesture.Rotate events are combined to prevent triggering handlers too often
- Apply code formatting

hansmbakker committed:
Refactoring
- Change NuimoBluetoothController constructor signature to take deviceId instead of BluetoothLEDevice so that it can recreate _bluetoothLEDevice after retrying to connect (since it is disposed when disconnecting)
- Simplify the retrieval of services and characteristics
- Tune the gesture event throttling

hansmbakker committed:
- Added new touch event (touch & long touch)
- Resharper code cleanup

hansmbakker committed:
Add built in led matrices

hansmbakker committed:
- Add heartbeat, rebootToDfu, fly sensor calibration, model and hardware functions
- Update led matrix characteristic

SeidChr reviewed (commented):
Wow, something is going on here :) 
Looking forward to migrate the addons to my API

larsblumberg commented on NuimoSDK/NuimoGestureEvent.cs:
```diff
@@ -21,10 +21,30 @@ public enum NuimoGesture
+        Unknown,
+        Black,
+        //TODO White or Silver?
+        Silver
```
We call them _white_ :)

hansmbakker replied:
Renamed to White, also added Gold from https://github.com/AravinthPanch/nuimo-bluetooth-low-energy-python

larsblumberg commented on NuimoSDK/INuimoController.cs:
```diff
@@ -5,35 +5,48 @@ namespace NuimoSDK
+        event Action<INuimoController, NuimoColor> ColorRead;
+        event Action<INuimoController> LedMatrixDisplayed;
+        event Action<INuimoController, int> BatteryPercentageChanged;
+         event Action<INuimoController, object> HeartbeatReceived;
```
One leading space too much

hansmbakker replied:
Fixed

larsblumberg commented on NuimoSDK/NuimoGestureEvent.cs:
```diff
@@ -21,10 +21,30 @@ public enum NuimoGesture
         SwipeDown,
+        TouchLeft,
+        TouchRight,
+        //TouchTop seems not to work
```
Yes, we don't make it yet available since touch-top (de-)activates the fly sensor. The event is however reserved in case of future use.

hansmbakker replied:
Ok, I documented it with a "remarks" comment

larsblumberg commented on NuimoSDK/NuimoGestureEvent.cs:
```diff
@@ -21,10 +21,30 @@ public enum NuimoGesture
+        TouchBottom,
+        LongTouchLeft,
+        LongTouchRight,
+        //LongTouchTop seems not to work
```
Yes, for same reason as above

hansmbakker replied:
Also documented

larsblumberg commented on NuimoSDK/NuimoGestureEvent.cs:
```diff
@@ -21,10 +21,30 @@ public enum NuimoGesture
+        LongTouchBottom,
         FlyLeft,
         FlyRight,
+        //Removed from firmware?
```
Yes

hansmbakker replied:
Removed from code

larsblumberg commented on NuimoSDK/NuimoGestureEvent.cs:
```diff
@@ -21,10 +21,30 @@ public enum NuimoGesture
         FlyRight,
+        //Removed from firmware?
         FlyBackwards,
+        //Removed from firmware?
```
Yes

hansmbakker replied:
Removed from code

larsblumberg commented on NuimoSDK/NuimoLedMatrix.cs:
```diff
@@ -5,19 +5,23 @@ namespace NuimoSDK
+        public NuimoLedMatrix()
+        {
+        }
+
```
What's the reason behind having this empty constructor? The instantiated matrix will not be properly initialized, i.e. `Leds` will be empty.

hansmbakker replied:
That one is fixed now, too

larsblumberg commented on NuimoSDK/PairedNuimoManager.cs:
```diff
@@ -1,21 +1,62 @@
         }
     }
-}
+}
```
Great! We have at this point no time to test the Nuimo discovery on Windows, but it looks good!

larsblumberg reviewed (changes requested):
Great Hans, awesome PR! Thanks for taking your time to port the new features from our other SDKs to this Windows SDK. Due to time constraints we couldn't update this Windows SDK and we're very grateful for your contribution.

I have only a few points to discus, such as my question about the empty matrix constructor. Please see below.

hansmbakker committed:
Default value for Leds
Remove project.json
Target Windows build 15063

hansmbakker committed:
Refactor NuimoColor to separate file
Fix White and add Gold color from https://github.com/AravinthPanch/nuimo-bluetooth-low-energy-python
use deviceId instead of bluetoothLeDevice for Identifier because bluetoothLeDevice might be null
add documentation

hansmbakker committed:
Fix leading space

hansmbakker committed:
Move platform-independent files to NuimoSDK.Core project which is netstandard1.4.
This allows you to write your application logic in lightweight .net standard projects, while the platform-dependent bluetooth part can be in a platform-specific (UWP) project.

hansmbakker commented:
I moved the platform-independent files to NuimoSDK.Core project which targets .net standard1.4.

This allows you to write your application logic in lightweight .net standard projects, while the platform-dependent bluetooth part can be in a platform-specific (UWP) project. This might need an extra Nuget-package which the NuimoSDK package depends on.

larsblumberg commented:
Thanks! I'm merging your PR. For the future we can move `NuimoColor` into `INuimoController` itself as the color can be considered part of the controller definition, same as `NuimoConnectionState` and `NuimoLedMatrixWriteOptions`. Can we publish the library on nuget?

hansmbakker commented:
@larsblumberg Great, thanks for merging!
As for types in separate files: this is for maintainability so that you can see by the file name what content to expect - see http://stackoverflow.com/questions/2434990/one-class-per-file-rule-in-net

As for publishing on nuget: I think it can be published, I have some ideas for refactoring NuimoBluetoothController according to https://en.baydachnyy.com/2017/04/28/uwp-working-with-bluetooth-devices-part-1/ (using separate classes per Bluetooth characteristic) but that can go in a separate pull request.

Please note that there will be two dlls now, and you can decide whether to pack them in the same nuget package or to create two separate packages.

larsblumberg commented:
The most up-voted answer to [the question you're citing](), actually [suggests to have multiple classes per file](http://stackoverflow.com/a/2435040/543875) – if it makes sense. In our case it does and I pledge for keeping the `INuimoController` class together with its enums.

I think we should pack both DLLs into a single file.

Can you please elaborate again in a few sentences what the platform specific part of the Bluetooth API on various MS Windows platforms is? I thought MS is about to unify their Bluetooth API, aren't they?

hansmbakker commented:
Yes, I agree that that "one type per file" is a thing each time can agree on.

You could use [-IncludeReferencedProjects](https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package#additional-options) to put everything in one file.

hansmbakker commented:
This is a very long comment since I'm afraid you have a very broad question and I don't know exactly what you mean with platform (it can refer to phone vs desktop, or Win32 vs UWP for example) and what you mean with unifying them. Also I'm not an expert here, but I think Windows has several platform API groups:
- the Win32 APIs (used by applications called Win32 apps, Desktop apps etc and needed for traditional things like windows services)
- the UWP APIs (used by applications called UWP apps, Windows Store apps, etc, available from windows 10)
- the Windows 8 Store APIs (also Store apps, but from windows 8).

As I understand it, generally speaking the UWP APIs have more recent features but they are more restrictive (for example with UWP you only have file access to certain folders, you need to request permissions via the manifest, etc). But you need them if you want to publish your app in the Windows Store.

I have no experience with BLE on Win32, maybe @Vittel or @Inrego can comment on that.

With UWP there are some device platform-specific things to keep in mind when using e.g. Bluetooth devices to trigger a background task: https://docs.microsoft.com/en-us/windows/uwp/launch-resume/support-your-app-with-background-tasks.

With .Net somebody could be using Xamarin to implement the bluetooth part for Nuimo there separately for Android and iOS devices.

Finally, .Net Standard is gaining traction but that does not contain any bluetooth APIs.

Inrego commented:
The desktop APIs (which @wind-rider refers to as Win32 APIs), does not have any BLE functionality. It is however possible with a portable library project type to target Windows 10, Xamarin, UWP and more. That is what I did in this PR: https://github.com/getsenic/nuimo-windows/pull/13

Having this project type, allows the bluetooth api's that is normally only in UWP to be consumed in a desktop application through that portable library assembly.

I haven't looked at this PR in detail apart from reading the comments here. But from my understanding, I see no reason to split it in 2 projects/assemblies.

hansmbakker commented:
In general it is a good idea to keep the platform-independent stuff (INuimoController et al) separated from the platform-dependent stuff (Bluetooth implementation).

The advantage of that is, for example, that you can do something like http://stackoverflow.com/questions/43705801/reference-uwp-library-from-net-standard-class-library. The implementation of that you can see in https://github.com/wind-rider/nuimo-hub. Other than that, it gives everybody the chance of implementing / using INuimoController without being forced to make it an UWP project.

@Inrego you did that indeed by creating a PCL which was the way to go at the time you created #13. However, [PCL is a technology which is being superseded by .net standard](https://docs.microsoft.com/en-us/dotnet/articles/standard/library#comparison-to-portable-class-libraries). I wouldn't want to use it for new projects anymore. 

I believe having these two parts separated like explained above is the cleanest and most future-proof solution.

Inrego commented:
My point was that with PCL, there wasn't really any platform-specific code. The full library worked on all platforms.
I haven't worked with .NET Standard yet, but doesn't it offer the same possibilities?

hansmbakker commented:
In a .Net Standard project, the APIs that are available are
- the APIs that are defined in the version of .Net Standard that the project is targeting
- any optional APIs coming from Nuget packages or projects that you reference.

Bluetooth APIs are not part of .Net Standard I believe - for that you need to have a platform specific project.

In the table at https://docs.microsoft.com/en-us/dotnet/articles/standard/library#net-platforms-support you can see that if you want to reference a .Net Standard project in your UWP 10.0 project, the highest .Net Standard version you can use is 1.4.

As for PCLs - there you don't have version levels but PCL profiles. And I believe that, roughly spoken, the more platforms you target, the less APIs you can use because it makes available only the APIs that are available on all platforms you target. I don't know how Bluetooth APIs can be accessed from PCLs.

Inrego commented:
You are right that's how PCL's work. I don't remember which targets I had in my PR, but I remember it was most of them (don't think it targeted Silverlight or old versions of Windows Phone for example).

The bluetooth APIs were available.

However, the PCL profile did not target any version of Windows below 10, since they don't have the bluetooth APIs.
So I guess, there are some cases in which you split it up like that - where it would make sense.
For example, if there suddenly pops up a nuget package that implements BLE on Windows < 10, you could still use this library and just reimplement the bluetooth specific parts for that reference.

hansmbakker commented:
For Xamarin a quick google shows https://github.com/xabre/xamarin-bluetooth-le and https://github.com/aritchie/bluetoothle/blob/master/readme.md. So people could probably use them to implement Nuimo for Xamarin, but I haven't used them myself.

hansmbakker commented:
@Inrego according to https://docs.microsoft.com/en-us/windows/uwp/devices-sensors/bluetooth-dev-faq you can use these UWP APIs from Win32 programs as explained on https://blogs.windows.com/buildingapps/2017/01/25/calling-windows-10-apis-desktop-application/ but I haven't tried this myself.

Inrego commented:
Yes exactly, that's what I am pointing out.
